### PR TITLE
HDDS-1372. getContainerWithPipeline for a standalone pipeline fails with ConcurrentModificationException.

### PR DESCRIPTION
**Pull request by nandakumar131**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1191 | trunk passed |
| +1 | compile | 52 | trunk passed |
| +1 | checkstyle | 21 | trunk passed |
| +1 | mvnsite | 34 | trunk passed |
| +1 | shadedclient | 758 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 43 | trunk passed |
| +1 | javadoc | 22 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 35 | the patch passed |
| +1 | compile | 24 | the patch passed |
| +1 | javac | 24 | the patch passed |
| +1 | checkstyle | 15 | the patch passed |
| +1 | mvnsite | 27 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 789 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 45 | the patch passed |
| +1 | javadoc | 19 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 117 | server-scm in the patch passed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 3348 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-682/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/682 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux fca415acbe43 4.4.0-138-generic #164~14.04.1-Ubuntu SMP Fri Oct 5 08:56:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 8b6deeb |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-682/1/testReport/ |
| Max. process+thread count | 366 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-682/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java [hunk 1]
```diff
@@ -194,6 +197,49 @@ public void testGetContainerWithPipeline() throws Exception {
...
+    final AtomicBoolean runUpdaterThread =
+        new AtomicBoolean(true);
+
+    Thread updaterThread = new Thread(() -> {
+      while (runUpdaterThread.get()) {
+        try {
+          containerManager.removeContainerReplica(cId, cReplica);
+          containerManager.updateContainerReplica(cId, cReplica);
+        } catch (ContainerNotFoundException e) {
+          e.printStackTrace();
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: This should be changed to an Assert

Comment (mukul1987) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: This should be converted to an Assert.fail

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java [hunk 2]
```diff
@@ -194,6 +197,49 @@ public void testGetContainerWithPipeline() throws Exception {
...
+        } catch (ContainerReplicaNotFoundException e) {
+          Assert.fail("Missing Container Replica for " + id);
+        }
+      }
+    });
+
+    updaterThread.setDaemon(true);
+    updaterThread.start();
+
+    IntStream.range(0, 10).forEach(i -> {
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: Lets have a few more iterations to make sure we have a higher probability of hitting this race. How about a value of 1000 ?

Comment (nandakumar131) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/TestSCMContainerManager.java]: 1000 will be very high, let make it 100. I was able to reproduce the issue every time with just 10 iterations.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 24 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1015 | trunk passed |
| +1 | compile | 27 | trunk passed |
| +1 | checkstyle | 18 | trunk passed |
| +1 | mvnsite | 30 | trunk passed |
| +1 | shadedclient | 670 | branch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 51 | trunk passed |
| +1 | javadoc | 21 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 30 | the patch passed |
| +1 | compile | 23 | the patch passed |
| +1 | javac | 23 | the patch passed |
| +1 | checkstyle | 13 | the patch passed |
| +1 | mvnsite | 28 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 726 | patch has no errors when building and testing our client artifacts. |
| +1 | findbugs | 47 | the patch passed |
| +1 | javadoc | 17 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 105 | server-scm in the patch passed. |
| +1 | asflicense | 25 | The patch does not generate ASF License warnings. |
| | | 2935 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-682/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/682 |
| Optional Tests |  dupname  asflicense  compile  javac  javadoc  mvninstall  mvnsite  unit  shadedclient  findbugs  checkstyle  |
| uname | Linux c9c093e50fd4 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 215ffc7 |
| maven | version: Apache Maven 3.3.9 |
| Default Java | 1.8.0_191 |
| findbugs | v3.1.0-RC1 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-682/2/testReport/ |
| Max. process+thread count | 450 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-682/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request merged**